### ISSUE TITLE
framework api int64 compatible problem

### ISSUE THREAD

**Issue by mttbx** [tensorflow/tensorflow#29820]
It seems tensorflow have compatible problem for Shard and int64
```
auto job = [&](int64 start, int64 limit)
{
            for (int64 i = start; i<limit; i++)
            {
                std::cout << start << std::endl; // Segmentation fault (core dumped)
            }
};

auto job = [&](int64 start, int64 limit)
{
            for (int i=0; i<4; i++)
            {
                std::cout << start << std::endl; // ok
            }
};

const DeviceBase::CpuWorkerThreads& worker_threads = *(context->device()->tensorflow_cpu_worker_threads());
const int64_t shard_cost = 10000;
Shard(worker_threads.num_threads, worker_threads.workers,
              (int64)4, shard_cost, job);

```
system info:
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): linux ubuntu 18.04
- TensorFlow installed from (source or binary):binary
- TensorFlow version (use command below):1.13 v1.13.1-0-g6612da8951
- Python version:3.5
- CUDA/cuDNN version:10/7.5
- GPU model and memory:7.5/24gb

**Comment by achandraa** [comment]
Will it be possible to provide a full code snippet that we can use to reproduce the issue. This will help us to proceed faster. Thanks!

**Comment by mttbx** [comment]
sure
testbug.cc:
```
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/register_types.h"
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/framework/tensor_shape.h"
#include "tensorflow/core/framework/register_types.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/util/work_sharder.h"

#include <iostream>
#include <cmath>

using namespace tensorflow;

REGISTER_OP("TestBug")
    .Output("dummy: float");

class TestBugOp : public OpKernel
{
public:
explicit TestBugOp(OpKernelConstruction* context)
        : OpKernel(context)
{

}

void Compute(OpKernelContext* context) override
{
    Tensor* dummy = NULL;
    OP_REQUIRES_OK(context, context->allocate_output(0, {1},
                                                     &dummy));
    auto job = [&](int64 start, int64 limit)
    {
        for (int64 i = start; i<limit; i++)
        {
            std::cout << start << std::endl; // Segmentation fault (core dumped)
        }
    };
    
    /*
    auto job = [&](int64 start, int64 limit)
    {
        for (int i = 0; i<4; i++)
        {
            std::cout << i << std::endl; // ok
        }
    };
    */

    const DeviceBase::CpuWorkerThreads& worker_threads = *(context->device()->tensorflow_cpu_worker_threads());
    const int64_t shard_cost = 10000;
    Shard(worker_threads.num_threads, worker_threads.workers,
              (int64)100, shard_cost, job);

}
};

REGISTER_KERNEL_BUILDER(
  Name("TestBug").Device(DEVICE_CPU),
  TestBugOp
);
    
```
testbug.py:
```
import tensorflow as tf
extr_module = tf.load_op_library('./build/libextr_module.so')

dummy = extr_module.test_bug()

with tf.Session() as sess:
    sess.run([dummy])
```
CMakeLists.txt:
```
CMAKE_MINIMUM_REQUIRED(VERSION 2.8)
PROJECT(extr_module)


# compiler flags
SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -O2 ${OpenMP_CXX_FLAGS} -Wall -fPIC -D_GLIBCXX_USE_CXX11_ABI=0 -DGOOGLE_CUDA=1")

# TensorFlow dependencies
EXECUTE_PROCESS(COMMAND python3.5 -c "import os; os.environ['TF_CPP_MIN_LOG_LEVEL']='3'; import tensorflow as tf; print(tf.sysconfig.get_include(), end='', flush=True)"  OUTPUT_VARIABLE TF_INC)

EXECUTE_PROCESS(COMMAND python3.5 -c "import os; os.environ['TF_CPP_MIN_LOG_LEVEL']='3'; import tensorflow as tf; print(tf.sysconfig.get_lib(), end='', flush=True)"  OUTPUT_VARIABLE TF_LIB)


MESSAGE(STATUS "Found TF_INC: " ${TF_INC})
#MESSAGE(STATUS "Found TF_INC_EXTERNAL: " ${TF_INC}/external/nsync/public)
MESSAGE(STATUS "Found TF_LIB: " ${TF_LIB})


INCLUDE_DIRECTORIES(${TF_INC})
#INCLUDE_DIRECTORIES(${TF_INC}/external/nsync/public)
LINK_DIRECTORIES(${TF_LIB})


ADD_LIBRARY(extr_module SHARED
  testbug.cc
)

TARGET_LINK_LIBRARIES(extr_module tensorflow_framework)
```
test:
```
mkdir build
cd build
cmake ..
make
cd ..
python3 testbug.py
```

**Comment by mttbx** [comment]
testbug.cc:
```

#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/register_types.h"
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/framework/tensor_shape.h"
#include "tensorflow/core/framework/register_types.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/util/work_sharder.h"

#include <iostream>
#include <cmath>

using namespace tensorflow;

REGISTER_OP("TestBug")
    .Attr("attr: int")
    .Output("dummy: float")
    .SetShapeFn([](::tensorflow::shape_inference::InferenceContext* c) {
        int32_t attr;
        TF_RETURN_IF_ERROR(c->GetAttr("attr", &attr));
    });

class TestBugOp : public OpKernel
{
public:
explicit TestBugOp(OpKernelConstruction* context)
        : OpKernel(context)
{

}

void Compute(OpKernelContext* context) override
{
    Tensor* dummy = NULL;
    OP_REQUIRES_OK(context, context->allocate_output(0, {1},
                                                     &dummy));
}
};

REGISTER_KERNEL_BUILDER(
  Name("TestBug").Device(DEVICE_CPU),
  TestBugOp
);
```
    
testbug.py:
```

import tensorflow as tf
extr_module = tf.load_op_library('./build/libextr_module.so')

dummy = extr_module.test_bug()

with tf.Session() as sess:
    sess.run([dummy])
```
CMakeLists.txt:
```

CMAKE_MINIMUM_REQUIRED(VERSION 2.8)
PROJECT(extr_module)


# compiler flags
SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -O2 ${OpenMP_CXX_FLAGS} -Wall -fPIC -D_GLIBCXX_USE_CXX11_ABI=0 -DGOOGLE_CUDA=1")

# TensorFlow dependencies
EXECUTE_PROCESS(COMMAND python3 -c "import os; os.environ['TF_CPP_MIN_LOG_LEVEL']='3'; import tensorflow as tf; print(tf.sysconfig.get_include(), end='', flush=True)"  OUTPUT_VARIABLE TF_INC)

EXECUTE_PROCESS(COMMAND python3 -c "import os; os.environ['TF_CPP_MIN_LOG_LEVEL']='3'; import tensorflow as tf; print(tf.sysconfig.get_lib(), end='', flush=True)"  OUTPUT_VARIABLE TF_LIB)


MESSAGE(STATUS "Found TF_INC: " ${TF_INC})
#MESSAGE(STATUS "Found TF_INC_EXTERNAL: " ${TF_INC}/external/nsync/public)
MESSAGE(STATUS "Found TF_LIB: " ${TF_LIB})


INCLUDE_DIRECTORIES(${TF_INC})
#INCLUDE_DIRECTORIES(${TF_INC}/external/nsync/public)
LINK_DIRECTORIES(${TF_LIB})


ADD_LIBRARY(extr_module SHARED
  testbug.cc
)

TARGET_LINK_LIBRARIES(extr_module tensorflow_framework)
```
test:
```

mkdir build
cd build
cmake ..
make
cd ..
python3 testbug.py
```

**Comment by mttbx** [comment]
problem solved, I changed g++ to 4.8 version, and all bug solved. solution is referenced to https://github.com/tensorflow/tensorflow/issues/13308
